File: starfyre/store.py
import uuid

import js

from starfyre.component import Component

from .dom_methods import render

store = {}
observers: dict[uuid.UUID, list[Component]] = {}


def create_signal(initial_state=None):
    """Create a signal to be used in a component."""
    global store
    id = uuid.uuid4()

    def getState(element):
        print("This is the type", type(element))
        if id in observers:
            observers[id].append(element)
        else:
            observers[id] = [element]

        return store.get(id, initial_state)

    def setState(state):
        store[id] = state
        for component in observers[id]:
            print("observer component", component)
            if component and component.parentDom:
                parentDom = component.parentDom
                parentDom.removeChild(component.dom)
            else:
                parentDom = js.document.getElementById("root")
                parentDom.innerHTML = ""

            print("rendering", component)
            # should be done in batching
            render(component)

    return [getState, setState]

File: starfyre/js/store.js
observers = {}; // uuid , list[observers]

const domIdMap = {};

function addDomIdToMap(domId, pyml) {
  domIdMap[domId] = pyml;
}

function getPymlFromDomId(domId) {
  return domIdMap[domId];
}

function render(domId) {
  const element = document.getElementById(domId);
  const pyml = getPymlFromDomId(domId);
  element.innerHTML = `${eval(pyml)}`;
}

function create_signal(initial_state) {
  let id = Math.random() * 1000000000000000; // simulating uuid
  let state = initial_state;

  function use_signal(domId) {
    if (!domId) {
      // when the domId is not provided, it means that the signal is used in a function component
      return store[id] || initial_state;
    }

    if (observers[id]) {
      observers[id].push(domId);
    } else {
      observers[id] = [domId];
    }

    return state;
  }

  function set_signal(newState) {
    state = newState;

    if (!observers[id]) {
      return;
    }

    observers[id].forEach((element) => {
      render(element);
    });
  }

  function get_signal() {
    return state;
  }

  return [use_signal, set_signal, get_signal];
}
